[RFC] treewide: convert mtd-mac-address to generic implementation

Ansuel opened the pull request:
This pr is to simplify the conversion from mtd-mac-address to nvmem implementation.
The address read function is kept but the increment functions are generalized and applied from any source (nvmem or mtd)

An example of the nvmem conversion for a target (ipq806x) is provided to better show what is needed to make this transition.

One question: Anyone know why the of_add_property is used in the mtd function? It's not used in the nvmem function? 
Don't know if that can cause some problem with the mac-address-increment generalization.

s-2 commented:
Awesome, the various ways of setting mac addresses are quite a hassle at the moment, great to see efforts towards migrating to `nvmem-cells`!

One reason that had me use base-files scripts in the past (rather than dts) were the different formats used by vendors for storing mac addresses; If I understand correctly, upstream nvmem-cells currently supports only binary mac addresses of 6 bytes, e.g.
`reg = <0x0 0x6>;`

However for some devices it could also be 12 (ascii hex) or 17 bytes (ascii hex with separator, mostly `:` or `-`), thus base-files scripts currently support three different formats: `mtd_get_mac_binary` / `mtd_get_mac_ascii` / `mtd_get_mac_text`.
I hope one day we could convice upstream maintainers to accept a more flexible parser directly within nvmem, though I'm not quite sure about how much of this should happen automagically (e.g. based on the size implied by the reg attribute?), maybe even EUI-64 length might be relevant one day.

Also for pcie wifi devices, when there is a valid-looking bogus mac address at `<&art 0x1002>`, this can be fixed either in `10_fix_wifi_mac`, or by not declaring mtd-caldata in the dts and instead patching it within `10-ath9k-eeprom` / `11-ath10k-caldata`. (Setting both caldata and mac in the dts is currently impossible for devices that use the 12/17-bytes ascii/text mac format)

For some devices, ethernet mac is handled by `10_fix_eth_mac.sh` in preinit, which exists in addition to what can be done within `02_network`...

So, it is all quite confusing (especially for new developers) and thus great to see some progress here :)

Regarding actual dts style, I would prefer to create a separate node for nvmem-cells, that could be referenced by the particular firmware-partition (assuming there is usually only one partition that contains any data of interest).

Maybe in the long term, also the `label-mac-device` alias could be superseded / complemented with an alias that directly refers to the nvmem-cells node, rather than a node that is an actual network device.

s-2 commented:
> But consider that the increment and all this kind of fun thing won't be accepted upstream ever. As oem implement various of special way to handle mac-address, the kernel won't support all of them, so they will never support any.

I see, so there is probably not much use for anything else than fixed binary values in upstream nvmem-cells by now.

Regarding the increment, many devices I have seen actually have multiple mac addresses defined in `art` etc., but sometimes people seem to prefer extracting only the first one and solve the rest with increment, maybe since that looks easier?
Not sure about those particular ipq devices though, but maybe for a few of the older devices that currently use increment, the remaining mac addresses (for wifi / 5 GHz) might actually be stored somewhere in the flash as well...


> About this, I will post a patch to add support for ath10k to use the of_get_address function to get the mac-address

👍 


> &defaultmac {
> 	compatible = "nvmem-cells";
> 
> 	macaddr_gmac: macaddr_gmac1@8 {
> 		reg = <0x8 0x6>;
> 	};
> }

The dts node for flash partitions is usually quite deep already (and also long for ipq806x), I think this really looks better in terms of structure 🙂 Besides I've seen vendors (mostly D-Link ath79) that re-use the same nvmem layout for multiple devices from a series, so these declarations could be put into dtsi, even when the mtd layout itself is different.

ynezz commented:
> But consider that the increment and all this kind of fun thing won't be accepted upstream ever.

This is probably just a wrong interpretation of the @robherring [review comment](https://lore.kernel.org/linux-mtd/CAL_JsqKhyeh2=pJcpBKkh+s3FM__DY+VoYSYJLRUErrujTLn9A@mail.gmail.com/). My understanding is, that it wasn't rejected, it just needs to be handled differently.

Ansuel commented:
Comments here gave me an idea. Something like a nvmem parser that modify
the readed value.

Il Mer 31 Mar 2021, 15:09 Petr Štetiar ***@***.***> ha
scritto:

> But consider that the increment and all this kind of fun thing won't be
> accepted upstream ever.
>
> This is probably just a wrong interpretation of the @robherring
> <https://github.com/robherring> review comment
> ***@***.***/>.
> My understanding is, that it wasn't rejected, it just needs to be handled
> differently.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/4041#issuecomment-811054390>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQWLWULPNCAQ56AL45DTGMNINANCNFSM42C74KWA>
> .
>

s-2 commented:
> @robherring [review comment](https://lore.kernel.org/linux-mtd/CAL_JsqKhyeh2=pJcpBKkh+s3FM__DY+VoYSYJLRUErrujTLn9A@mail.gmail.com/).

How is "stuffing the dts using standard properties" to be interpreted, i.e. some platform-specific code would overwrite the values to `/proc/device-tree` during boot, so that drivers could retrieve them in a generic way? This would need to happen at least in pre-init then.

Ansuel commented:
I updated the ipq806x commit using the suggested structure. It actually looks better and more organized.
Thx for the suggestion @s-2

s-2 commented:
> mhhh, the best we can do it this (for c2600 for example)

> I updated the ipq806x commit using the suggested structure. It actually looks better and more organized.

I only just realized that `compatible = "nvmem-cells";` is not a node of its own within the mtd partition, but actually just properties declared on the same level, so it is indeed impossible(?) to declare this within a .dtsi file (in case of multiple devices with different mtd layouts sharing the same nvmem structure). And I would not like to use C-style macros in the dts here, so your solution is definitely preferred.

hnyman commented:
> Someone should make a list of the needed changes,

One MAC address change type is "locally administered bit" toggling that some routers like wndr3800 do. I proposed that functionality with #1257 but it was rejected at that time.

adschm commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8062-wg2600hp3.dts:
```diff
@@ -399,6 +399,18 @@
+&factory {
+	compatible = "nvmem-cells";
+
+	macaddr_gmac1: macaddr_gmac1@0 {
```
Solely cosmetic, but do the nodes need unique names?

Since we have DT labels in any case, maybe something like this would be enough:

`macaddr_gmac1: macaddr@0 {`

?

Ansuel replied:
I need to check if the nvmem framework like that.

adschm replied:
Note that this is only an open suggestion, since I do not really think the duplication is necessary here.

Ansuel replied:
no problem for nvmem so i will use the generic macaddr@....

adschm commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-ad7200-c2600.dtsi:
```diff
@@ -280,6 +280,14 @@
+	compatible = "nvmem-cells";
+
+	macaddr_gmac: macaddr_gmac1@8 {
+			reg = <0x8 0x6>;
```
Wrong indent.

adschm commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-d7800.dts:
```diff
@@ -280,6 +280,18 @@
+
+	macaddr_gmac2: macaddr_gmac2@0 {
+		reg = <0x0 0x6>;
+	};
```
Hmm, I'd personally prefer ordering based on the offset here.

BTW, this whole "naming after the interface" will fail when the same address location is used several times (for different interfaces).
I'm not sure how to nicely name these cases, though.

Ansuel replied:
We can consider to drop the interface naming and call them mac1 mac2.... What do you think?

adschm replied:
Being generic wouldn't be the worst approach either, yes. But which is first? If you just follow the flash locations, it might conflict with the numbers used on a label etc.
I'm still thinking whether we would find a nice approach to include the offset/location, as that's probably the only information actually "belonging" to the address itself.

Ansuel replied:
(i'm preparing a script to try and check if the conversion can be automated...)
anyway with your suggestion, i'm currently using 

macaddr_art_0 for example (macaddr_*partition name*_*reg offset*

adschm replied:
Something like that, yes. I still consider it a bit ugly and wonder whether we can find a more appealing alternative.
In any way, if there is only one partition containing the addresses, one could omit the partition name, of course (macaddr_0 or macaddr_0x0).

Ansuel replied:
Isn't that too generic? I think a bit of redundancy can be acceptable to directly understand what is the phandle about.

adschm commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-vr2600v.dts:
```diff
@@ -305,6 +305,14 @@
+	compatible = "nvmem-cells";
+
+	macaddr_gmac: macaddr_gmac1@0 {
+			reg = <0x0 0x6>;
```
broken indent.

adschm commented:
Looks like a step forward. Thanks.

Regarding label-mac-device, if this is possible, I'm all open to replacing it with label-mac-address.
Eventually, we only care about the address anyway, and if ambiguous, the device to link to is chosen arbitrarily anyway.

Ansuel committed:
ipq806x: convert mtd-mac-address to nvmem implementation

Define nvmem-cells and convert mtd-mac-address to nvmem implementation.
The conversion is done with an automated script.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
lantiq: convert mtd-mac-address to nvmem implementation

Define nvmem-cells and convert mtd-mac-address to nvmem implementation.
The conversion is done with an automated script.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
mvebu: convert mtd-mac-address to nvmem implementation

Define nvmem-cells and convert mtd-mac-address to nvmem implementation.
The conversion is done with an automated script.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ath79: convert mtd-mac-address to nvmem implementation

Define nvmem-cells and convert mtd-mac-address to nvmem implementation.
The conversion is done with an automated script.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ramips: convert mtd-mac-address to nvmem implementation

Define nvmem-cells and convert mtd-mac-address to nvmem implementation.
The conversion is done with an automated script.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
octeon: convert mtd-mac-address to nvmem implementation

Define nvmem-cells and convert mtd-mac-address to nvmem implementation.
The conversion is done with an automated script.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
mpc85xx: convert mtd-mac-address to nvmem implementation

Define nvmem-cells and convert mtd-mac-address to nvmem implementation.
The conversion is done with an automated script.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
mediatek: convert mtd-mac-address to nvmem implementation

Define nvmem-cells and convert mtd-mac-address to nvmem implementation.
The conversion is done with an automated script.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel committed:
ipq40xx: convert mtd-mac-address to nvmem implementation

Define nvmem-cells and convert mtd-mac-address to nvmem implementation.
The conversion is done with an automated script.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel commented:
@adschm this will be fun ahahaha
Anyway the automated script worked without error... problem is now review alle these scripts...
What do you think of the new structure?

Ansuel commented:
> Regarding label-mac-device, if this is possible, I'm all open to replacing it with label-mac-address.
> Eventually, we only care about the address anyway, and if ambiguous, the device to link to has been chosen arbitrarily anyway.

I didn't understand this. Can you rephrase this?

Ansuel committed:
kernel: add ath10k support for of_get_mac_address

ath10k doesn't currently support the standard function to get mac-address from the dts.
Add this for both ath10k and ath10k-ct

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel commented:
(I also added the ath10k support patch, it looks like ath9k and other ethernet driver already supports this)

oldium commented on target/linux/ath79/dts/ar7161_netgear_wndr.dtsi:
```diff
@@ -222,7 +225,30 @@
+		reg = <0x0 0x6>;
+	};
+
+	macaddr_art_00: macaddr@00 {
```
It looks like this is here twice.

Ansuel replied:
Nice find! any idea why there is 0x0 and 0x00 ? Nobody notice that?

hnyman replied:
> any idea why there is 0x0 and 0x00 ?

No special reason. Looks like I have used different notation in b58df9919 (originally 25ea21d7) and 437e35f8 

WNDR3700v1,v2 and 3800 has three MAC stored there.
0x00 is used for eth0 and wlan0,
0x06 is used for eth1
0x0c is used for wlan1

The representation of 0 differs between lines 156 and 212. 0x0 vs 0x00

Ansuel replied:
i'm updating the script to use the 0x0 notation and I will push fixed dts. Think this problem is only in ath79, am i wrong?

hnyman replied:
There might be lots of variation if you look closer. Especially the ar71xx/ath79 device conversions have been made with so many people, and partially by modifying/copying from various other files, that there may well be other similar things. (also between DTSI and DTS).

Ansuel replied:
Assuming 0x00 variants are only present in ath79, i should have fixed all the dts... (there were 9)

Ansuel commented:
Rebase with latest master.
I updated the mediatek commit and the ath79 commit with the 0x00 variant fixed.

RussellSenior commented:
@Ansuel is the script published anywhere?

Ansuel commented:
@RussellSenior Should i post the script?

Also should I update the pr? Any news about this?

ynezz commented:
@Ansuel Please rebase one more time and I'll hopefully merge it tomorrow.

adschm commented on target/linux/ath79/dts/qca9558_allnet_all-wap02860ac.dts:
```diff
@@ -90,5 +90,6 @@
 	mtd-mac-address = <&art 0x0>;
-	mtd-mac-address-increment = <1>;
+	mac-address-increment = <1>;
+	qca-no-eeprom;
```
Unrelated?

adschm commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-r7500v2.dts:
```diff
@@ -303,10 +304,24 @@
 	};
-};
\ No newline at end of file
+};
```
Add empty line ...

adschm commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-wxr-2533dhp.dts:
```diff
@@ -447,4 +449,17 @@
 	};
-};
\ No newline at end of file
+};
```
Missing empty line.

adschm commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-r7800.dts:
```diff
@@ -442,8 +445,23 @@
+&art {
+	compatible = "nvmem-cells";
+	#address-cells = <1>;
+	#size-cells = <1>;
```
Being a child node of the partitions that have the same setup, wouldn't the address-/size-cells be inherited, so we could omit them here (in general)?

Ansuel replied:
I need to check but i remember that the buildroot complained with missing address and size without this

adschm commented:
Typos in commit title: rampis -> ramips, oceteon -> octeon, ipq4019->ipq40xx

adschm commented:
In the long term, I wonder whether having a separate node for this (as you did) or integrating it directly in the partition table would be the better option. But for the moment I'd consider this OT, it should not delay the merge by any means ...

Ansuel commented:
@ynezz I pass my script to every target and in theory all should be good....
Anyway since it was asked i'm using this 2 script (homemade sh*t so don't judge ahahhahaha) (tried to find a good python module to handle dts but they are compatible with compiled dts...)

https://gist.github.com/Ansuel/ed2c574fef60842ef7d7183952fdeb9a

Ansuel commented:
@adschm i'm honestly using the separate node to not complicate too much the migration script... If we really want i can consider tweaking the script

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

ynezz commented:
[Seems](https://gitlab.com/ynezz/openwrt/-/pipelines/332333483) like, that at least [ramips](https://gitlab.com/ynezz/openwrt/-/jobs/1401740194) is having built issues.

john-tho commented:
> [Seems](https://gitlab.com/ynezz/openwrt/-/pipelines/332333483) like, that at least [ramips](https://gitlab.com/ynezz/openwrt/-/jobs/1401740194) is having build issues.



[`Error: ../dts/mt7621_dlink_dir-8xx.dtsi:137.1-9 Label or path factory not found`](https://gitlab.com/ynezz/openwrt/-/jobs/1401740194#L1203)

Perhaps because the root label path reference [`target/linux/ramips/dts/mt7621_dlink_dir-8xx.dtsi` `&factory {`](https://github.com/Ansuel/openwrt/blob/724b40d0ea24dd4f537ad3d82eba67b021962dbc/target/linux/ramips/dts/mt7621_dlink_dir-8xx.dtsi#L137) is defined after it is used (in a file included later in the dts). Maybe move this node into:
[`target/linux/ramips/dts/mt7621_dlink_flash-16m-a1.dtsi`](https://github.com/openwrt/openwrt/blob/b0483b19f981bbe024cd447cfe469765f49cc587/target/linux/ramips/dts/mt7621_dlink_flash-16m-a1.dtsi#L28) and [`target/linux/ramips/dts/mt7621_dlink_flash-16m-r1.dtsi`](https://github.com/openwrt/openwrt/blob/b0483b19f981bbe024cd447cfe469765f49cc587/target/linux/ramips/dts/mt7621_dlink_flash-16m-r1.dtsi#L28)? I am not sure though. I cannot find if for device-tree or dtc, referenced root labels need to be defined first?

<details><summary>log parts for first error</summary>

```
mipsel-openwrt-linux-musl-cpp -nostdinc -x assembler-with-cpp -I/builds/ynezz/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/linux-5.10.46/arch/mips/boot/dts -I/builds/ynezz/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/linux-5.10.46/arch/mips/boot/dts/include -I/builds/ynezz/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/linux-5.10.46/include/ -undef -D__DTS__  -o /builds/ynezz/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/image-mt7621_dlink_dir-867-a1.dtb.tmp ../dts/mt7621_dlink_dir-867-a1.dts

/builds/ynezz/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/linux-5.10.46/scripts/dtc/dtc -O dtb -i../dts/ -Wno-unit_address_vs_reg -Wno-simple_bus_reg -Wno-unit_address_format -Wno-pci_bridge -Wno-pci_device_bus_num -Wno-pci_device_reg -Wno-avoid_unnecessary_addr_size -Wno-alias_paths -Wno-graph_child_address -Wno-graph_port -Wno-unique_unit_address   -o /builds/ynezz/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/image-mt7621_dlink_dir-867-a1.dtb /builds/ynezz/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/image-mt7621_dlink_dir-867-a1.dtb.tmp

Error: ../dts/mt7621_dlink_dir-8xx.dtsi:137.1-9 Label or path factory not found
FATAL ERROR: Syntax error parsing input tree

make[5]: *** [Makefile:207: /builds/ynezz/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/image-mt7621_dlink_dir-867-a1.dtb] Error 1
```
</details>

Ansuel commented:
> [Seems](https://gitlab.com/ynezz/openwrt/-/pipelines/332333483) like, that at least [ramips/mt7621](https://gitlab.com/ynezz/openwrt/-/jobs/1401740194) and [ath79/generic](https://gitlab.com/ynezz/openwrt/-/jobs/1401740195) are having build issues. Can you perhaps try to build/compile all the DTSes with the SDK?

(sorry... anyway cloning your gitlab repo and use that should be also good? my pc is melting and would take lots of time to test all targets)

ynezz commented:
> my pc is melting and would take lots of time to test all targets

No problem, just let me know once the issues are fixed and I'll re-run the build tests.

Ansuel commented:
i'm checking the error from your buildroot...
about rampis are we sure this change is correct?
https://github.com/openwrt/openwrt/commit/2001c0ca9fd2bf95827e23ba867fa06fc3b0a3e7#

in the dtsi `mt7621_dlink_dir-8xx.dtsi` `&factory` is referenced 
but is only defined by `mt7621_dlink_flash-16m-a1.dtsi` that is included after
(think this is the error that cause the factory not defined, as 8xx is defined before factory 

@adschm I was thinking... shouldn't it better to split the switch and pcie part that use factory tag from the 8xx.dtsi and put it in a separate one and included after the factory node is defined?

adschm commented:
> @adschm I was thinking... shouldn't it better to split the switch and pcie part that use factory tag from the 8xx.dtsi and put it in a separate one and included after the factory node is defined?

> art is referenced in qca9558_comfast_cf-wr650ac.dtsi but only defined in the dts. Isn't this wrong?

So, we finally have a reason to enforce this....

I'm not really sure what's the best solution to address this in the named cases. I'd like to have some time to make up my mind there.
IMO the easiest option would be to remove the relevant (few) cases from this patchset, and address them separately later. There are other devices that have been merged in the meantime with the old setup, so there will be additional cleanup patches anyway.

Ansuel commented:
@adschm i still have to setup a ci to test all the target with the dts changes as requested from yneez
So the idea is to not convert these strange dts and wait for a major cleanup 

Anyway i think tomorrow will setup the ci and i will polish this. I also notice there is a not related modification in 19c9c433f76bfdf0aadd0f74bf0f19842c42a9d8

Ansuel committed:
treewide: convert mtd-mac-address-increment* to generic implementation

Rework patch 681-NET-add-mtd-mac-address-support to implement
only the function to read the mac-address from mtd.
Generalize mtd-mac-address-increment function so it can be applied
to any source of of_get_mac_address.
Rename any mtd-mac-address-increment to mac-address-increment.
Rename any mtd-mac-address-increment-byte to mac-address-increment-byte.

This should make simplify the conversion of target to nvmem implementation.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

Ansuel commented:
@adschm i cleared the commits but still i have to build test them....

Do you have time to give this a check ?

ynezz commented:
> @adschm i cleared the commits but still i have to build test them....
> 
> Do you have time to give this a check ?

I've just build test them again https://gitlab.com/ynezz/openwrt/-/pipelines/338161615/failures

Ansuel commented:
ath79 seems to have failed.... think i have to will have to local test that... times has come to burn down my pc ahahah

ynezz commented:
```
../dts/mt7621.dtsi:503.13-507.8: ERROR (phandle_references): /ethernet@1e100000/mdio-bus/switch@1f/ports/port@0: Reference to non-existent node or label "macaddr_factory_4"
  also defined at ../dts/mt7621_iodata_wn-ax1167gr.dts:137.10-143.5
```

Ansuel commented:
yep notice the same

ynezz commented:
> think i have to will have to local test that... times has come to burn down my pc ahahah

You can probably just download the corresponding SDKs and run those dtb and dtc (dtc is included in the SDK) steps manually:

```
mipsel-openwrt-linux-musl-cpp -nostdinc -x assembler-with-cpp -I/builds/ynezz/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/linux-5.10.46/arch/mips/boot/dts -I/builds/ynezz/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/linux-5.10.46/arch/mips/boot/dts/include -I/builds/ynezz/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/linux-5.10.46/include/ -undef -D__DTS__  -o /builds/ynezz/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/image-mt7621_iodata_wn-ax1167gr2.dtb.tmp ../dts/mt7621_iodata_wn-ax1167gr2.dts

/builds/ynezz/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/linux-5.10.46/scripts/dtc/dtc -O dtb -i../dts/ -Wno-unit_address_vs_reg -Wno-simple_bus_reg -Wno-unit_address_format -Wno-pci_bridge -Wno-pci_device_bus_num -Wno-pci_device_reg -Wno-avoid_unnecessary_addr_size -Wno-alias_paths -Wno-graph_child_address -Wno-graph_port -Wno-unique_unit_address   -o /builds/ynezz/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/image-mt7621_iodata_wn-ax1167gr2.dtb /builds/ynezz/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7621/image-mt7621_iodata_wn-ax1167gr2.dtb.tmp
```

ynezz commented:
BTW I've just noticed, that I've had `CONFIG_TESTING_KERNEL=y` in my CI config, which means, that those build failures are comming from 5.10 kernel (those issues should be probably fixed as well), so building now without this option, thus with 5.4 kernel https://gitlab.com/ynezz/openwrt/-/pipelines/338248950

Ansuel commented:
ok it seems it doesn't take too much time so i think i will just compile locally...

Ansuel commented:
Ok ath79 fixed (compile tested...)
now i need to check the others

Ansuel commented:
@ynezz i should have tested all the target (can you run your ci one last time?)
@adschm here is the list of the dts that still use the old binding...
As you can see they are all dtsi and i notice there are also some bmips and 2 patch... (should i also update that? will try that...)

Anyway i can consider update the migration script to put these node in the dts that include the dtsi?
Did you find another way to fix this issue?
```
./lantiq/files/arch/mips/boot/dts/lantiq/vr9_avm_fritz736x.dtsi
./ath79/dts/qca9533_qca_ap143.dtsi
./ath79/dts/qca9563_tplink_tl-wpa8630p-v2.dtsi
./ath79/dts/ar9331_glinet_64xx.dtsi
./ath79/dts/qca9563_tplink_archer-x6-v2.dtsi
./ath79/dts/ar7241_netgear_wnr2200.dtsi
./ath79/dts/qca9563_tplink_re450.dtsi
./ath79/dts/qca9561_tplink_archer-c5x.dtsi
./ath79/dts/ar7241_ubnt_unifi.dtsi
./ath79/dts/qca9563_tplink_tl-wr1043n.dtsi
./ath79/dts/qca9561_tplink_archer-c6x.dtsi
./ath79/dts/qca9563_tplink_archer-x7-v5.dtsi
./ath79/dts/qca9563_elecom_wrc-ghbk2-i.dtsi
./ath79/dts/ar9331_tplink_tl-wr710n.dtsi
./ath79/dts/qca9558_comfast_cf-wr650ac.dtsi
./ath79/dts/qca9558_tplink_archer-c.dtsi
./ath79/dts/ar7161_netgear_wndr.dtsi
./ath79/dts/ar9341_tplink.dtsi
./ramips/dts/rt5350_poray_m4.dtsi
./ramips/dts/mt7628an_unielec_u7628-01.dtsi
./ramips/dts/mt7621_unielec_u7621-06.dtsi
./ramips/dts/mt7620a_zbtlink_zbt-we826.dtsi
./ramips/dts/rt5350_vocore_vocore.dtsi
./ramips/dts/rt5350_7links_px-4885.dtsi
./ramips/dts/mt7628an_widora_neo.dtsi
./ramips/dts/rt3050_alphanetworks_asl26555.dtsi
./ramips/dts/mt7620n_nexx_wt3020.dtsi
./ramips/dts/rt5350_nexx_wt1520.dtsi
./ramips/dts/mt7621_unielec_u7621-01.dtsi
./ramips/dts/mt7621_ubnt_unifi.dtsi
./ramips/dts/mt7620a_netgear_ex2700_wn3000rp-v3.dtsi
./ramips/dts/mt7621_dlink_dir-8xx.dtsi
./ramips/dts/mt7621_netgear_sercomm_ayx.dtsi
./ramips/dts/rt5350_asiarf_awm002-evb.dtsi
./ramips/dts/mt7621_iodata_wn-ax1167gr.dts
./ramips/dts/rt5350_nixcore_x1.dtsi
./ramips/dts/mt7628an_vocore_vocore2.dtsi
./ramips/dts/rt3050_allnet_all0256n.dtsi
./bmips/dts/bcm63168-comtrend-vr-3032u.dts
./bmips/dts/bcm6362-netgear-dgnd3700-v2.dts
./bmips/dts/bcm6318-comtrend-ar-5315u.dts
./bmips/dts/bcm6328-comtrend-ar-5387un.dts
./bmips/dts/bcm6368-comtrend-vr-3025u.dts
./bmips/dts/bcm6358-huawei-hg556a-b.dts
./mediatek/patches-5.10/101-dts-update-mt7629-rfb.patch
./mediatek/patches-5.4/0005-dts-mt7629-add-gsw.patch
```

Ansuel committed:
bmips: convert mtd-mac-address to nvmem implementation

Define nvmem-cells and convert mtd-mac-address to nvmem implementation.
The conversion is done with an automated script.

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

ynezz commented:
> can you run your ci one last time

https://gitlab.com/ynezz/openwrt/-/pipelines/339295676

ynezz commented:
> > can you run your ci one last time
> 
> https://gitlab.com/ynezz/openwrt/-/pipelines/339295676

CI seems happy, so I've pushed it, thanks!

ynezz commented:
ath25 [seems to be broken](https://buildbot.openwrt.org/master/images/#/builders/46/builds/194/steps/47/logs/stdio)

```make
  MODPOST 93 modules
ERROR: "of_platform_device_create" [/builder/shared-workdir/build/build_dir/target-mips_mips32_musl/linux-ath25/backports-5.10.42-1/drivers/net/wireless/ath/ath10k/ath10k_core.ko] undefined!
OpenWrt-libtool: link: mips-openwrt-linux-musl-gcc -Wall -Waggregate-return -Wmissing-declarations -Wmissing-prototypes -Wredundant-decls -Wshadow -Wstrict-prototypes -Wlogical-op -Winline -pipe -DALL_INCLUSIVE 
make[7]: *** [scripts/Makefile.modpost:94: __modpost] Error 1
```

Ansuel commented:
ath25 doesn't have CONFIG_OF enabled?

Ansuel commented:
@ynezz should we put an ifdef in the ath10k patch to disable the `of_platform_device_create` if not available ?

ynezz commented:
> @ynezz should we put an ifdef in the ath10k patch to disable the `of_platform_device_create` if not available ?

I don't see any other solution for that.

Ansuel commented:
Quick fix https://github.com/openwrt/openwrt/pull/4380

ynezz commented:
> JFY, this series of commits broke Gl.Inet mt-1300 beryl(mt7621) - I've had to revert it in my local build. So basically the problem is that device cannot read it's mac from partition and every reboot it generates a new one.

https://lists.infradead.org/pipermail/openwrt-devel/2021-July/035870.html

Ansuel commented:
I'm investigating that... The mtk ethernet driver use the normal API that also mtd-mac-address. I need to check if I found the cause of the problem without asking for debugging.

ynezz commented:
> I'm investigating that... The mtk ethernet driver use the normal API that also mtd-mac-address. I need to check if I found the cause of the problem without asking for debugging.

It's probably just missing `CONFIG_NVMEM=y` ?

fblaese commented:
> https://lists.infradead.org/pipermail/openwrt-devel/2021-July/035870.html

I'm having the same issue with the Xiaomi Mi Router 4A (100m) (ramips/mt76x8).
Including `CONFIG_NVMEM=y` does fix the issue for me.

However, there is yet another issue: This change seems to break `get_mac_label_dt`, because the `mac-address` parameter isn't available in `/sys/firmware/devicetree` anymore, which is used here:
https://github.com/openwrt/openwrt/blob/c3e252d08032de80e7c82759cc9763a310a97144/package/base-files/files/lib/functions/system.sh#L25-L26

Ansuel commented:
Ok it's the nvmem flag not compiled. Do we have space problem on ramips?

Also what is the use of that function ? (If we really need that we can add
a patch that register the mac-address when the generic function is called
(that was present in the mtd-mac-address patch but I thought it wasn't
really used)

Il Mar 20 Lug 2021, 13:38 Fabian Bläse ***@***.***> ha
scritto:

> https://lists.infradead.org/pipermail/openwrt-devel/2021-July/035870.html
>
> I'm having the same issue with the Xiaomi Mi Router 4A (100m)
> (ramips/mt76x8).
> Including CONFIG_NVMEM=y does fix the issue for me.
>
> However, there is yet another issue: This change seems to break
> get_mac_label_dt, because the mac-address parameter isn't available in
> /sys/firmware/devicetree anymore, which is used here:
>
> https://github.com/openwrt/openwrt/blob/c3e252d08032de80e7c82759cc9763a310a97144/package/base-files/files/lib/functions/system.sh#L25-L26
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/4041#issuecomment-883322801>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQSQSO7QDQNRFNT4G5LTYVN2BANCNFSM42C74KWA>
> .
>

fblaese commented:
> Also what is the use of that function

This function is used to get the mac address printed on the device's label. [[1]](https://openwrt.org/docs/guide-developer/mac.address#using_the_label_mac_address)
When reading this information from the device tree (`get_mac_label_dt`), the `label-mac-device` alias is used to identify the device tree node.

Ansuel commented:
@ynezz i notice 'label-mac-device' is widely used to get a base mac-address... currently we point to the the gmac node most of the time
I still have to check if we use the increment function
For now i see 2 solution
- Introduce another function that with some magic find a way to access the nvmem data without editing the dts
- Change the dts to point to the nvmem cell and enable CONFIG_NVMEM_SYSFS that in theory should give direct read for the nvmem cell (i still have to check that... in the nvmem documentation they document that the cell should be readable but for some reason there isn't any read entry in the sysfs dir, probably caused by CONFIG_NVMEM_SYSFS disabled?)

adschm commented:
So nobody thought about label-mac-device and it's broken now? ............................

Note that we don't care at all about the _device_ here, it's the address we want. So if you find a way to link directly to the nvmem nodes and replace label-mac-device by label-mac-addr(ess) in the DTS (and adjust the shell function accordingly), that's actually an improvement. Moving the information away from DTS again would definitely be a huge step backwards ...

Ansuel commented:
@adschm
I would find the reason we don't have the read entry in the nvmem and add a check in the function that also try to get the info from the cell... (and change every dts that use the nvmem framework to point directly to the nvmem cell)
That should be the cleanest way to fix this.

Any sorry if I ask but what is the actual use of the function?

blocktrron commented:
@Ansuel I have a patch ready which fixes the label-mac it by adding the property to the node. This is hacky, I'll admit that. However, we don't have to introduce a hastily built, more complex solution. It's self contained to a separate patch, so I'd consider this fine-enough.

I also add the missing Config for ramips

Ansuel commented:
Can you post the patch so I can see how hacky it is?

blocktrron commented:
See my staging tree

Ansuel commented:
Ok I checked the patch and actually doesn't look that bad. We can consider
a cleanup later I think

Il Mar 20 Lug 2021, 22:53 David Bauer ***@***.***> ha scritto:

> See my staging tree
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/4041#issuecomment-883700875>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQTIJHRTBCBACQONXIDTYXO3LANCNFSM42C74KWA>
> .
>

john-tho commented:
Another possible alternative for fixing `label-mac-device`?

extend [`get_mac_label_dt`](https://github.com/openwrt/openwrt/blob/master/package/base-files/files/lib/functions/system.sh#L18) with a means to get the macaddr from the driver for the /sys/class/net device that matches the label-mac-device devicetree alias

I don't know a simpler method to get the /sys/class device of a devicetree node in the shell.
Untested, and could surely be tidier:

```diff
diff --git a/package/base-files/files/lib/functions/system.sh b/package/base-files/files/lib/functions/system.sh
index 80e417182a..162c844507 100644
--- a/package/base-files/files/lib/functions/system.sh
+++ b/package/base-files/files/lib/functions/system.sh
@@ -25,6 +25,24 @@ get_mac_label_dt() {
 	macaddr=$(get_mac_binary "$basepath/$macdevice/mac-address" 0 2>/dev/null)
 	[ -n "$macaddr" ] || macaddr=$(get_mac_binary "$basepath/$macdevice/local-mac-address" 0 2>/dev/null)
 
+	# retrieve mac address from driver
+	if [ ! -n "$macaddr" ]; then
+		local net_device
+		local of_node
+
+		# test label-mac-device alias against network device of_node
+		for net_device in /sys/class/net/*/of_node; do
+			of_node=$(readlink -f "$net_device")
+			of_node="${of_node#/sys/firmware/devicetree/base/}"
+			net_device="${net_device%/of_node}"
+
+			if [ "$macdevice" = "/$of_node" ]; then
+				macaddr=$(cat "$net_device/address")
+				break
+			fi
+		done
+	fi
+
 	echo $macaddr
 }
 

```

Cheers

adschm commented:
> And sorry if I ask but what is the actual use of the function?

In many large networks (particularly Freifunk communities) devices are identified by their MAC addresses or an IPv6 address built from it. Since a device typically has multiple addresses, it is convenient to choose the one which is printed directly on the device (which means, the label). So, the function essentially generates a unique identifier for each device which the user can know by looking at the box.
Since this information (the label MAC address) is definitely device-based - i.e. describing the device - it makes sense to provide it directly in OpenWrt, though we are not immediately using it. For the same reason, it should be in the DTS if possible.

Only remotely related to the actual problem, but FYI:
https://openwrt.org/docs/guide-developer/mac.address#label_mac_address

ynezz commented:
>     * Change the dts to point to the nvmem cell and enable CONFIG_NVMEM_SYSFS that in theory should give direct read for the nvmem cell (i still have to check that... in the nvmem documentation they document that the cell should be readable but for some reason there isn't any read entry in the sysfs dir, probably caused by CONFIG_NVMEM_SYSFS disabled?)

I can't check it on the device right now, but IIUC this would be complex, as the sysfs exposes complete nvmem memory, not just the MAC address part, so you would still need to duplicate at least the information about the offset in the userspace.

> I have a patch ready which fixes the label-mac it by adding the property to the node. This is hacky, I'll admit that. However, we don't have to introduce a hastily built, more complex solution. It's self contained to a separate patch, so I'd consider this fine-enough.

+1

> I also add the missing Config for ramips

Thanks, I've just pushed it along with the same fix for bmips.

ivanich commented:
@ynezz Weird, built ramips image with https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=373c08bcbb08b49715919597ed400abb4dd169c9 but mac address still random

Ansuel commented:
@ynezz random theory... as we still have random mac with nvmem enabled... and assumed the fact that the driver used the same api (`of_get_mac_address`)
I wonder if the bootloader is doing some magic and putting some bad value in the mac-address...

I notice that in the old mtd patch, `of_get_mac_address_mtd` was called BEFORE the other propriety check... 
Wonder if that is the problem... 
if that's the case it could be that
- the bootloader insert a bad (empty) mac address as a propriety
- the api found the propriety before the mtd and nvmem check and so they are skipped
- the driver reject the invalid mac addr and generate a random one

I see the change in priority order the only big difference between the old and the new implementation (referring to the changes done to the api)

Ansuel commented:
Anyway some update on this...
With some debugging we found out that 
`of_find_device_by_node` is having some problem on finding the pdev... Any hint on how we can use a substitute to get the dev connected to the device node?

ivanich commented:
Wouldn't it be better to revert your changes for devic
es which affected with this instead of keeping  master in the unusable state for those devices?

Ansuel commented:
Yes that is also a way.
But honestly it's really a Mac problem the device is usable just random Mac
Anyway in theory I found a solution I need to check for the wan. But yes
reverting the changes for the mediatek and rampis target is not a bad idea
if it's actually a problem.

Il Gio 22 Lug 2021, 23:56 Ivanich ***@***.***> ha scritto:

> Wouldn't it be better to revert your changes for devic
> es which affected with this instead of keeping master in the unusable
> state for those devices?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/4041#issuecomment-885261167>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQQSOI6Q4RDHIY6IR7TTZCHZ5ANCNFSM42C74KWA>
> .
>

Ansuel commented:
Still waiting for testing but in theory this should fix the mac on mediatek based devices
https://github.com/openwrt/openwrt/pull/4391

Ansuel commented:
Actually asking upstream it seems they actually fixed the problem with nvmem...
`f10843e04a07  of: net: fix of_get_mac_addr_nvmem() for non-platform devices`

So i will have to backport that and this way we should be able to drop any workaround needed...

Ansuel commented:
https://github.com/openwrt/openwrt/pull/4392
If anyone have time, pls test this a give me some feedback

s-2 commented:
As a side note, I was wondering whether upstream has any intentions of having the `nvmem-cells` driver actually _provide_ data contents itself in the long term, rather than declaring their posititon, i.e. we still need to have the individual driver resolve the `phandle` property and figure out how to retrieve contents. For OpenWrt, this is probably not relevant since the provider is always mtd.
After all, I was just curious whether we could update `base-files` scripts to retrieve mac addresses directly from sysfs, but there is no direct way to read contents from within a shell script:

```
root@OpenWrt:/sys/firmware/devicetree/base/ahb/spi@1f000000/flash@0/partitions/partition@ff0000/macaddr@1002# ll
drwxr-xr-x    2 root     root             0 Jul 27 05:03 ./
drwxr-xr-x    3 root     root             0 Jul 27 05:03 ../
-r--r--r--    1 root     root             8 Jul 27 05:03 name
-r--r--r--    1 root     root             4 Jul 27 05:03 phandle
-r--r--r--    1 root     root             8 Jul 27 05:03 reg
```

Ansuel commented:
I'm doing some experiment adding support for this. Do you think this can be useful for openwrt?

s-2 commented:
> I'm doing some experiment adding support for this. Do you think this can be useful for openwrt?

Honestly I don't have much practical experience with targets other than ath79 or ramips, but it seems that ethernet mapping and mac address assignment are still the main reason for basically *every* device to require some boilerplate handling code within `02_network`; e.g. I just checked my other PR for changes required due to nvmem-cells, but it turned out mac addresses could not even be set in dts for switch ports... but regarding `ath79`, I would not touch anything in terms of switch ports for the moment, until the target is migrated to DSA.

On the other hand, looking at e.g. `ramips/dts/mt7621_dlink_dir-xx60-a1.dtsi` even with DSA it can be confusing (WAN mac is set directly for `port@4` by overriding LAN mac which was set via the `gmac0` node).

So, the philosophical question would be, do we want to assign
`physical port -> interface (LAN/WAN/...) -> mac address`
or
`interface (LAN/WAN/...) <- physical port -> mac address`
currently we seem to have a mixuture of both approaches, and it may be argued that WAN/LAN is something virtual, to be configured by the user, not related to the board (despite maybe socket plastic color or port labelling), so it should not be in dts, which shall only describe the ~~hardware~~ parts of the board not altered by the user?
`nvmem-cells` however describe the positions of important values in the flash, which could be considered board-related, just as the device label denoting the mac address.

So technically, physical ports and LAN/WAN mac addresses would belong into dts, but the assignment of ports to interfaces should be configureable by the user (except for the intial defaults, matching oem behaviour), and if the user changes port assignment, this should not require any redundant configuration (e.g. overwrting mac adress via uci etc.) 🤔 

Sorry for making this unneccessarily complicated, sometimes I just get pedantic when thinking too much about simple things 😂

adschm commented:
@Ansuel @ynezz @blocktrron 
I will push patches to remove/consolidate the remaining cases of mtd-mac-address today.
We would thus be able to remove mtd-mac-address[-increment] support entirely after that. If somebody wants to prepare patches, feel free to do so ...

Ansuel commented:
@adschm news on this? Should i make a quick search to check how much device we need to change to actually drop mtd-mac-address ?

adschm commented:
I think I dropped all remaining uses of mtd-mac-address[-increment] now (I grepped for "mtd-mac" in target/linux ...).

If you like, please prepare a PR for dropping the patch(es), so we can discuss there ...

Ansuel commented:
@adschm i prepared #4467
Did I miss something?

ecsv commented:
Seems like this PR broke mac addresses for OpenMesh (and similar devices which get their partition table supplied via  bootargs' mtdparts from u-boot).

The nvmem driver is no longer trying to get the correct mtd via its name but relies on the partition being created by the devicetree (and not cmdline's mtdparts).
